File: app/src/main/java/com/randomappsinc/studentpicker/Misc/JSONUtils.java
package com.randomappsinc.studentpicker.Misc;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by alexanderchiou on 1/31/16.
 */
public class JSONUtils {
    public static final String NAMES_KEY = "names";
    public static final String ALREADY_CHOSEN_NAMES_KEY = "alreadyChosenNames";

    // Given a list of names, converts it to a JSON and stringifies it
    public static String serializeNameList(List<String> names, List<String> alreadyChosenNames) {
        try {
            JSONObject nameListJson = new JSONObject();

            // Store current names in list
            JSONArray namesArray = new JSONArray();
            for (String name : names) {
                namesArray.put(name);
            }
            nameListJson.put(NAMES_KEY, namesArray);

            // Store already chosen names
            JSONArray alreadyChosenNamesArray = new JSONArray();
            for (String alreadyChosenName : alreadyChosenNames) {
                alreadyChosenNamesArray.put(alreadyChosenName);
            }
            nameListJson.put(ALREADY_CHOSEN_NAMES_KEY, alreadyChosenNamesArray);

            return nameListJson.toString();
        }
        catch (JSONException e) {
            return "";
        }
    }

    // Given a serialized JSON "cache" string of a name list, extracts the names
    public static List<String> extractNames(String cachedList) {
        List<String> names = new ArrayList<>();
        try {
            JSONObject nameListJson = new JSONObject(cachedList);
            JSONArray namesArray = nameListJson.getJSONArray(NAMES_KEY);
            for (int i = 0; i < namesArray.length(); i++) {
                names.add(namesArray.getString(i));
            }
        }
        catch (JSONException ignored) {}
        return names;
    }

    // Given a serialized JSON "cache" string of a name list, extracts the names
    public static List<String> extractAlreadyChosenNames(String cachedList) {
        List<String> alreadyChosenNames = new ArrayList<>();
        try {
            JSONObject nameListJson = new JSONObject(cachedList);
            JSONArray alreadyChosenNamesArray = nameListJson.getJSONArray(ALREADY_CHOSEN_NAMES_KEY);
            for (int i = 0; i < alreadyChosenNamesArray.length(); i++) {
                alreadyChosenNames.add(alreadyChosenNamesArray.getString(i));
            }
        }
        catch (JSONException ignored) {}
        return alreadyChosenNames;
    }
}

File: app/src/main/java/com/randomappsinc/studentpicker/utils/JSONUtils.kt
package com.randomappsinc.studentpicker.utils

import android.text.TextUtils
import com.randomappsinc.studentpicker.models.ListDO
import com.randomappsinc.studentpicker.models.NameDO
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.*

object JSONUtils {

    private const val NAME_LIST = "nameList"
    private const val LIST_NAME = "listName"
    private const val NAME = "name"
    private const val AMOUNT = "amount"

    @JvmStatic
    fun namesArrayToJsonString(chosenNames: List<String>): String {
        val alreadyChosenNamesArray = JSONArray()
        for (alreadyChosenName in chosenNames) {
            alreadyChosenNamesArray.put(alreadyChosenName)
        }
        return alreadyChosenNamesArray.toString()
    }

    @JvmStatic
    fun extractNamesHistory(namesArrayText: String?): List<String> {
        if (TextUtils.isEmpty(namesArrayText)) {
            return emptyList()
        }
        val alreadyChosenNames: MutableList<String> = ArrayList()
        try {
            val namesArray = JSONArray(namesArrayText)
            for (i in 0 until namesArray.length()) {
                alreadyChosenNames.add(namesArray.getString(i))
            }
        } catch (ignored: JSONException) {
        }
        return alreadyChosenNames
    }

    fun serializeNameList(listDOs: List<ListDO>): String {
        val nameListArray = JSONArray()
        for (listDO in listDOs) {
            val listDOJson: JSONObject? = createNameListJson(listDO)
            if (listDOJson != null) {
                nameListArray.put(listDOJson)
            }
        }
        return nameListArray.toString()
    }

    private fun createNameListJson(listDO: ListDO): JSONObject? {
        try {
            val nameListJsonObject = JSONObject().apply {
                put(LIST_NAME, listDO.name)
            }

            val names = JSONArray()
            for (nameDO in listDO.namesInList) {
                val nameJson = JSONObject().apply {
                    put(NAME, nameDO.name)
                    put(AMOUNT, nameDO.amount)
                }
                names.put(nameJson)
            }
            nameListJsonObject.put(NAME_LIST, names)

            return nameListJsonObject
        } catch (e: JSONException) {
            return null
        }
    }

    fun deserializeNameListsJson(nameListsJson: String?): List<ListDO> {
        val listDOs = ArrayList<ListDO>()
        try {
            val nameListsArray = JSONArray(nameListsJson)
            for (i in 0 until nameListsArray.length()) {
                val nameListJson = nameListsArray.getJSONObject(i)
                listDOs.add(getNameListFromJson(nameListJson))
            }
        } catch (ignored: JSONException) {}
        return listDOs
    }

    private fun getNameListFromJson(nameListJson: JSONObject): ListDO {
        val listDO = ListDO()
        try {
            listDO.name = nameListJson.getString(LIST_NAME)

            val nameDOs = ArrayList<NameDO>()
            val nameList = nameListJson.getJSONArray(NAME_LIST)
            for (i in 0 until nameList.length()) {
                val nameJson = nameList.getJSONObject(i)
                val nameDo = NameDO().apply {
                    name = nameJson.getString(NAME)
                    amount = nameJson.getInt(AMOUNT)
                }
                nameDOs.add(nameDo)
            }
            listDO.namesInList = nameDOs
        }
        catch (ignored: JSONException) {}
        return listDO
    }
}